### ISSUE TITLE
set declarative_base() / registry.declarative_base() as legacy, introduce new base class approach

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#7562]
the `declarative_base()` function was a convenient approach at the time it was created, however it is not in line with current Python practices.     The comments on https://github.com/python/typing/discussions/1020 indicate that typing tools have no plans to support this pattern in any way right now, and I also get the vibe, based on the comments, as well as that [the actual mypy hook for dynamic classes refers to our own API as the example](https://github.com/python/mypy/blob/fa16759dc37fa313a2a607974d51d161c8ba5ea6/mypy/plugin.py#L474), that we are the only ones doing this in any prominent way.

on top of that, the declarative_base()'s reliance on metaclasses, which are usually unnecessary these days now that we have `__init_subclass__()`, has always been a burden on users who were often trying to use their own metaclasses, as well as that it caused problems we had to fix with end-user classes that used `__init_subclass__()` or `__class_getitem__()`.

on top of that we already have many other ways to mark classes as mapped, most conveniently the `@registry.mapped` decorator.

However, having a `Base` class is still a very handy pattern, I still find myself preferring it, and the whole world uses it.  We'd also like to get rid of the metaclass requirement without breaking backwards compatibility.

Turns out that with the current API and `__init_subclass__()`, making a new declarative base that isn't too much trouble is stupidly easy, though it does require two lines on the part of the end user rather than one to create, like this:

```py

class DeclarativeBase(MappedBase):
    registry: ClassVar["registry"]
    metadata: ClassVar[MetaData]
    __mapper__: ClassVar[Mapper]
    __table__: Optional[FromClause]

    if typing.TYPE_CHECKING:
        def __init__(self, **kw: Any):
            ...

    def __init_subclass__(cls) -> None:
        if DeclarativeBase in cls.__bases__:
            cls.registry = registry()
            cls.metadata = cls.registry.metadata
        else:
            cls.registry.mapped(cls)
```

so end user code does:

```py
from sqlalchemy.orm import DeclarativeBase

class Base(DeclarativeBase):
    pass
```

that's your Base.  mypy no longer complains and you also have a nice place to put custom mixin logic (that works right now too).

so far there's two things we change in the API and we have basically everything we want, *except* for fully-typed init:

1. mapped_column constructor
2. new declarative base as above

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**WIP for ORM typing** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3495

**Comment by jvanasco** [comment]
We should compile a list of docs that need to be changed.

**Comment by CaselIT** [comment]
> We should compile a list of docs that need to be changed.

Great idea!

**Comment by zzzeek** [comment]
this is implemented, first revision can be seen at ff1ab665cb1

docs for all the new declarative patterns will be one big job done separately.
